File: src/solutions/2023/08/ts/main.ts
import { argv } from 'node:process';
import type { SolutionOptions } from 'src/options.ts';
import { run } from 'src/run.ts';

const options: SolutionOptions = {
   hasAlternate: true,
   hasIo: false,
};

function part1(input: string, _isTest: boolean): string {
   const sides = input.split('\n\n');
   const instructions = sides[0].replaceAll('L', '0').replaceAll('R', '1');
   const maps = sides[1]
      .split('\n')
      .map((str) => {
         const [dest, lr] = str.split(' = ');
         return [dest, lr.slice(1, lr.length - 1).split(', ')] as [string, [string, string]];
      })
      .reduce((pv, v) => {
         pv[v[0]] = v[1];
         return pv;
      }, {} as Record<string, [string, string]>);

   let i = 0;
   let nav = 'AAA';
   while (true) {
      const map = maps[nav];
      nav = map[instructions[i % instructions.length] as '0'];
      i++;
      if (nav === 'ZZZ') break;
   }
   return i.toString();
}

function gcd(a: number, b: number): number {
   return !b ? a : gcd(b, a % b);
}

function lcm(a: number, b: number): number {
   return (a * b) / gcd(a, b);
}

function part2(input: string, _isTest: boolean): string {
   const sides = input.split('\n\n');
   const instructions = sides[0].replaceAll('L', '0').replaceAll('R', '1');
   const maps = sides[1]
      .split('\n')
      .map((str) => {
         const [dest, lr] = str.split(' = ');
         return [dest, lr.slice(1, lr.length - 1).split(', ')] as [string, [string, string]];
      })
      .reduce((pv, v) => {
         pv[v[0]] = v[1];
         return pv;
      }, {} as Record<string, [string, string]>);

   const navs = Object.keys(maps).filter((s) => s[2] === 'A');
   const dists = new Array(navs.length).fill(0);
   for (let i = 0; i < navs.length; i++) {
      let j = 0;
      while (true) {
         const map = maps[navs[i]];
         navs[i] = map[instructions[j % instructions.length] as '0'];
         j++;
         if (navs[i][2] === 'Z') break;
      }
      dists[i] = j;
   }
   return dists.reduce((pv, v) => lcm(pv, v), 1).toString();
}

if (import.meta.main) {
   run(argv, part1, part2, options);
}

File: src/solutions/2023/08/ocaml/main.ml
let options = Hashtbl.create 128;;

Hashtbl.add options "has_alternate" true;;
Hashtbl.add options "has_io" false;;

let ( +/ ) = Int64.add
let ( */ ) = Int64.mul
let ( // ) = Int64.div

let rec gcd (a : int64) (b : int64) =
  if b = Int64.of_int 0 then a else gcd b (Int64.rem a b)
;;

let lcm a b = a */ b // gcd a b

let part_1 (input : string) (_is_test : bool) : string =
  let lines = input |> String.split_on_char '\n' in
  let instructions =
    lines
    |> List.hd
    |> String.to_seq
    |> Seq.map (fun x -> if x = 'L' then 0 else 1)
    |> Array.of_seq
  in
  let maps = Hashtbl.create 128 in
  lines
  |> List.tl
  |> List.filter (( <> ) "")
  |> List.iter (fun x ->
    let parts = String.split_on_char '=' x in
    let dest = List.nth parts 0 |> String.trim in
    let lr = List.nth parts 1 |> String.split_on_char ',' |> List.map String.trim in
    Hashtbl.add
      maps
      dest
      [| String.sub (List.nth lr 0) 1 3; String.sub (List.nth lr 1) 0 3 |]);
  let rec loop (i : int64) (nav : string) : int64 =
    let m = Hashtbl.find maps nav in
    let idx = Int64.to_int i mod Array.length instructions in
    let next = m.(instructions.(idx)) in
    let nexti = i +/ 1L in
    if next = "ZZZ" then nexti else loop nexti next
  in
  loop 0L "AAA" |> Int64.to_string
;;

let part_2 (input : string) (_is_test : bool) : string =
  let lines = input |> String.split_on_char '\n' in
  let instructions =
    lines
    |> List.hd
    |> String.to_seq
    |> Seq.map (fun x -> if x = 'L' then 0 else 1)
    |> Array.of_seq
  in
  let maps = Hashtbl.create 128 in
  let navs = ref [||] in
  lines
  |> List.tl
  |> List.filter (( <> ) "")
  |> List.iter (fun x ->
    let parts = String.split_on_char '=' x in
    let dest = List.nth parts 0 |> String.trim in
    let lr = List.nth parts 1 |> String.split_on_char ',' |> List.map String.trim in
    Hashtbl.add
      maps
      dest
      [| String.sub (List.nth lr 0) 1 3; String.sub (List.nth lr 1) 0 3 |];
    if String.get dest 2 = 'A' then navs := Array.append !navs [| dest |] else ());
  let rec loop (res : int64) (i : int) : int64 =
    if i = Array.length !navs
    then res
    else (
      let rec whileloop (j : int) : int =
        let m = Hashtbl.find maps !navs.(i) in
        let idx = j mod Array.length instructions in
        !navs.(i) <- m.(instructions.(idx));
        let nextj = j + 1 in
        if String.get !navs.(i) 2 = 'Z' then nextj else whileloop nextj
      in
      loop (lcm res (Int64.of_int (whileloop 0))) (i + 1))
  in
  loop 1L 0 |> Int64.to_string
;;

let () = Aoc.run Sys.argv part_1 part_2 options
